### ISSUE TITLE
Reconstructions look washed out

### ISSUE THREAD

**Issue by JeremyIV** [Nikolai10/PerCo#10]
I am using your pretrained models from Google Drive for this. I ran a command like the following
```
python src/compression_utils.py --model_path "models/cmvl_2024_full_train_.0313bpp" evaluate_ds --mode 0 --in_dir kodak-dataset/ --out_dir res/data/Kodak-original-.0313bpp
```
The resulting reconstructions are systematically washed-out looking: whites are not fully white, blacks are not fully black. For example:
![image](https://github.com/user-attachments/assets/b7e00203-077c-4453-b56f-80f42ea1a16b)
![image](https://github.com/user-attachments/assets/db9b093e-3757-4d97-8bf3-fb5a17ca91ca)

BTW I commented out line 79 of compression_utils.py to preserve the aspect ratio instead of squishing the kodak images:

```
    #img = TR.functional.resize(img, (512, 512), Image.BICUBIC)
```

But I see the same effect whether or not this line is commented.

Any thoughts on why? I know I have encountered similar problems when I wasn't properly normalizing the VAE latents during encoding/decoding. Just a guess :shrug: .

**Comment by Nikolai10** [comment]
Hello @JeremyIV,

thanks for sharing your observations.

**Normalizing VAE latents**
During [training](https://github.com/Nikolai10/PerCo/blob/master/src/train_sd_perco.py#L155), we use: `latents = latents * pipeline.vae.config.scaling_factor`.

During [inference](https://github.com/Nikolai10/PerCo/blob/master/src/pipeline_sd_perco.py#L1097), we use: `image = self.vae.decode(latents / self.vae.config.scaling_factor, return_dict=False, generator=generator)[0]`,

which seems fine.

**Colour preservation at higher bit-rates**

This is a reconstruction from our model at 0.1250 target rate:

![kodim04_0 1250](https://github.com/user-attachments/assets/1fa1946c-776d-463f-a48d-47f528a42568)

which looks much closer to the original image, also suggesting that there is no inherent bug.

**What is causing these issues at 0.0313bpp**

I suspect that at low bit rates, the hyper-encoder may prioritize storing semantic structure and edges over colour information, as it cannot afford to encode both effectively.

Hope this helps!
Nikolai

**Comment by JeremyIV** [comment]
Thank you for your quick response! 

So on the Kodak dataset I am getting an average PSNR of 
- 15.26 db with the 0.0019bpp model
- 19.40 db with the 0.0313 model
- 22.82 db with the 0.1250bpp model.

Can you confirm that this matches your expectations? Everything is working as intended?

**Comment by Nikolai10** [comment]
Please have a look at https://arxiv.org/pdf/2409.20255 (Fig. 5) for reference.

For a general discussion on PSNR, please consider:

> In Appendix B.1 we also report PSNR, but find these low-level distortion metrics to be less meaningful for low rates, 
see Figure 15.

(source: https://arxiv.org/pdf/2310.10325)

**Comment by JeremyIV** [comment]
Ok so here's the results I got when running your compression script:

```
perco_df = pd.DataFrame(data={
    "bpp": [.00363, 0.03239, 0.12614],
    "psnr": [15.26186, 19.39848, 22.82144],
    "lpips": [0.54597, 0.30008, 0.14269],
})
```

It's a little difficult to eyeball, but I think these measurements do look roughly on track with your reported results in Figure 5.

Thank you for your quick responses, I am closing the issue.

**Comment by JeremyIV** [comment]
> I suspect that at low bit rates, the hyper-encoder may prioritize storing semantic structure and edges over colour information, as it cannot afford to encode both effectively.

After running a quick experiment, I am not convinced by this argument. I can consistently increase the PSNR of the .0313 bpp reconstructions on the Kodak dataset by about 0.5 db by increasing the contrast of these images so they don't look so washed out. The rescaling that improves PSNR is roughly something like `reconstruction * 1.05 - 0.05` for images in the range [0,1]. 

This is so little information, just two low-precision scalars, that I don't buy the argument that the learned compression protocol just can't "afford" to communicate that information for a 0.5 db PSNR improvement. This performance doesn't seem like it's on any reasonable Pareto frontier, it seems like a bug.

**Comment by JeremyIV** [comment]
sorry for waffling, I'm not sufficiently convinced of my own argument here either that this needs to be an open issue. Closing.

**Comment by Nikolai10** [comment]
Hello @JeremyIV,

no problem, in case you find better hyper-parameters, please share them here :).

Ideally, you should evaluate your approach directly on the MSCOCO-30k dataset to obtain a larger quantitative comparison.
